File: amber/typescript/src/tracing.ts
import { AmberTransitionError, Provenance } from "./provenance.js";

export type TraceAttributes = Readonly<Record<string, string | number>>;

/** Return the default tracing attribute projection for provenance. */
export function toTraceAttributes(provenance: Provenance): TraceAttributes {
  if (!(provenance instanceof Provenance)) {
    throw new AmberTransitionError("trace value must be a Provenance instance");
  }
  provenance.validate();
  const attributes: Record<string, string | number> = {
    "amber.version": provenance.version,
    "amber.work_id": provenance.work_id,
    "amber.execution_id": provenance.execution_id,
    "amber.correlation_id": provenance.correlation_id,
    "amber.origin": provenance.origin,
    "amber.depth": provenance.depth,
    "amber.attempt": provenance.attempt,
    "amber.mode.kind": provenance.mode.kind,
  };
  const mode = provenance.mode;
  if (mode.kind === "retry" || mode.kind === "replay") {
    attributes["amber.mode.of_execution_id"] = mode.of_execution_id;
  }
  if (mode.kind === "replay") {
    attributes["amber.mode.replay_id"] = mode.replay_id;
  }
  const causation = provenance.causation;
  if (causation !== undefined) {
    attributes["amber.causation.kind"] = causation.kind;
    attributes["amber.causation.id"] = causation.id;
    if (causation.type !== undefined) {
      attributes["amber.causation.type"] = causation.type;
    }
  }
  return Object.freeze(attributes);
}

/** Return cloned trace attributes with Amber attributes applied. */
export function mergeTraceAttributes(
  existing: Readonly<Record<string, unknown>>,
  provenance: Provenance,
): Record<string, unknown> {
  return { ...existing, ...toTraceAttributes(provenance) };
}

File: amber/go/adapters/tracing/tracing.go
// Package ambertracing provides a tracing-library-neutral provenance projection.
package ambertracing

import amber "github.com/0xsj/ingen/amber"

// ToAttributes returns the default tracing attribute projection for provenance.
func ToAttributes(provenance amber.Provenance) (map[string]any, error) {
	if err := provenance.Validate(); err != nil {
		return nil, err
	}
	attributes := map[string]any{
		"amber.version":        provenance.Version(),
		"amber.work_id":        provenance.WorkID().String(),
		"amber.execution_id":   provenance.ExecutionID().String(),
		"amber.correlation_id": provenance.CorrelationID().String(),
		"amber.origin":         string(provenance.Origin()),
		"amber.depth":          provenance.Depth(),
		"amber.attempt":        provenance.Attempt(),
		"amber.mode.kind":      string(provenance.Mode().Kind),
	}

	mode := provenance.Mode()
	if mode.OfExecutionID != "" {
		attributes["amber.mode.of_execution_id"] = mode.OfExecutionID.String()
	}
	if mode.ReplayID != "" {
		attributes["amber.mode.replay_id"] = mode.ReplayID.String()
	}
	if causation, ok := provenance.Causation(); ok {
		attributes["amber.causation.kind"] = causation.Kind
		attributes["amber.causation.id"] = causation.ID.String()
		if causation.Type != "" {
			attributes["amber.causation.type"] = causation.Type
		}
	}
	return attributes, nil
}

// MergeAttributes returns a cloned attribute map with Amber attributes
// applied. The existing map is never mutated.
func MergeAttributes(existing map[string]any, provenance amber.Provenance) (map[string]any, error) {
	attributes, err := ToAttributes(provenance)
	if err != nil {
		return nil, err
	}
	merged := make(map[string]any, len(existing)+len(attributes))
	for key, value := range existing {
		merged[key] = value
	}
	for key, value := range attributes {
		merged[key] = value
	}
	return merged, nil
}
